// @swiftlysingh/excalidraw-cli/dist/index.js
/**
 * Excalidraw CLI - Programmatic API
 *
 * Create Excalidraw flowcharts programmatically.
 */
// Parser exports
export { parseDSL } from './parser/dsl-parser.js';
export { parseJSON, parseJSONString } from './parser/json-parser.js';
// Layout exports
export { layoutGraph } from './layout/elk-layout.js';
// Generator exports
export { generateExcalidraw, serializeExcalidraw } from './generator/excalidraw-generator.js';
// Factory exports (for advanced usage)
export { createNode, createArrow, createText } from './factory/index.js';
// Default options
export { DEFAULT_LAYOUT_OPTIONS } from './types/dsl.js';
export { DEFAULT_APP_STATE, DEFAULT_ELEMENT_STYLE } from './types/excalidraw.js';
/**
 * High-level API: Create an Excalidraw flowchart from DSL string
 */
export async function createFlowchartFromDSL(dsl) {
    const graph = (await import('./parser/dsl-parser.js')).parseDSL(dsl);
    const layoutedGraph = await (await import('./layout/elk-layout.js')).layoutGraph(graph);
    const excalidrawFile = (await import('./generator/excalidraw-generator.js')).generateExcalidraw(layoutedGraph);
    return (await import('./generator/excalidraw-generator.js')).serializeExcalidraw(excalidrawFile);
}
/**
 * High-level API: Create an Excalidraw flowchart from JSON input
 */
export async function createFlowchartFromJSON(input) {
    const graph = (await import('./parser/json-parser.js')).parseJSON(input);
    const layoutedGraph = await (await import('./layout/elk-layout.js')).layoutGraph(graph);
    const excalidrawFile = (await import('./generator/excalidraw-generator.js')).generateExcalidraw(layoutedGraph);
    return (await import('./generator/excalidraw-generator.js')).serializeExcalidraw(excalidrawFile);
}
