// @swiftlysingh/excalidraw-cli/dist/factory/node-factory.js
/**
 * Node Factory
 *
 * Creates Excalidraw shape elements (rectangle, diamond, ellipse).
 */
import { createBaseElement } from './element-factory.js';
/**
 * Map DSL style to Excalidraw properties
 */
function mapStyle(style) {
    if (!style)
        return {};
    return {
        strokeColor: style.strokeColor,
        backgroundColor: style.backgroundColor,
        strokeWidth: style.strokeWidth,
        strokeStyle: style.strokeStyle,
        fillStyle: style.fillStyle,
        opacity: style.opacity,
        roughness: style.roughness,
    };
}
/**
 * Create a rectangle element
 */
export function createRectangle(node, boundElements) {
    const styleProps = mapStyle(node.style);
    return {
        ...createBaseElement('rectangle', node.x, node.y, node.width, node.height, {
            id: node.id,
            roundness: { type: 3 }, // Adaptive roundness
            boundElements: boundElements || null,
            ...styleProps,
        }),
        type: 'rectangle',
    };
}
/**
 * Create a diamond element
 */
export function createDiamond(node, boundElements) {
    const styleProps = mapStyle(node.style);
    return {
        ...createBaseElement('diamond', node.x, node.y, node.width, node.height, {
            id: node.id,
            roundness: { type: 2 }, // Proportional roundness
            boundElements: boundElements || null,
            ...styleProps,
        }),
        type: 'diamond',
    };
}
/**
 * Create an ellipse element
 */
export function createEllipse(node, boundElements) {
    const styleProps = mapStyle(node.style);
    return {
        ...createBaseElement('ellipse', node.x, node.y, node.width, node.height, {
            id: node.id,
            roundness: null, // Ellipses don't use roundness
            boundElements: boundElements || null,
            ...styleProps,
        }),
        type: 'ellipse',
    };
}
/**
 * Create a node element based on type
 */
export function createNode(node, boundElements) {
    switch (node.type) {
        case 'diamond':
            return createDiamond(node, boundElements);
        case 'ellipse':
            return createEllipse(node, boundElements);
        case 'database':
            // Database is rendered as rectangle with special styling
            return createRectangle(node, boundElements);
        case 'rectangle':
        default:
            return createRectangle(node, boundElements);
    }
}
